// pully/dist/utils/speedometer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const human_1 = require("./human");
function currentTimestamp() {
    return Date.now() / 1000;
}
class Speedometer {
    constructor(windowSize = 15000) {
        this._start = { bytes: 0, timestamp: currentTimestamp() };
        this._history = [];
        this._window = windowSize / 1000;
    }
    get bytesPerSecond() {
        if (this._history.length === 0)
            return 0;
        const now = currentTimestamp();
        let oldestTimestamp = now - this._window;
        while (this._history.length > 0 && this._history[0].timestamp < oldestTimestamp)
            this._history.shift();
        const { older, newer } = this._getRange();
        return (newer.bytes - older.bytes) / (newer.timestamp - older.timestamp);
    }
    get currentSpeed() {
        return `${human_1.toHumanSize(this.bytesPerSecond)}/s`;
    }
    eta(targetBytes) {
        const currentSpeed = this.bytesPerSecond;
        const { newer } = this._getRange();
        const remainingBytes = targetBytes - newer.bytes;
        return remainingBytes / currentSpeed;
    }
    record(bytes) {
        this._history.push({
            timestamp: currentTimestamp(),
            bytes
        });
    }
    _getRange() {
        if (this._history.length === 0)
            return {
                older: {
                    timestamp: this._start.timestamp - 1,
                    bytes: 0
                },
                newer: this._start
            };
        let older = this._history[0];
        let newer = this._history[this._history.length - 1];
        if (this._history.length === 1) {
            older = this._start;
        }
        return { older, newer };
    }
}
exports.Speedometer = Speedometer;
