// pully-core/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const UrlPattern = require("url-pattern");
const sanitize = require('sanitize-filename');
const lodashTemplate = require('lodash.template');
const userPattern = new UrlPattern('/user/:username');
const channelPattern = new UrlPattern('/channel/:channelId');
function extractPlaylistId(url) {
    const urlData = url_1.parse(url, true, true);
    let playlistId = null;
    if (urlData.query.list) {
        playlistId = urlData.query.list;
    }
    return playlistId;
}
exports.extractPlaylistId = extractPlaylistId;
function extractVideoId(url) {
    const urlData = url_1.parse(url, true, true);
    let videoId = null;
    if (urlData.query.v) {
        videoId = urlData.query.v;
    }
    return videoId;
}
exports.extractVideoId = extractVideoId;
function extractChannelId(url) {
    const urlData = url_1.parse(url, true, true);
    let channelId = null;
    const channelData = channelPattern.match(urlData.pathname);
    if (channelData !== null && channelData.channelId) {
        channelId = channelData.channelId;
    }
    return channelId;
}
exports.extractChannelId = extractChannelId;
function extractUsername(url) {
    const urlData = url_1.parse(url, true, true);
    let username = null;
    const userData = userPattern.match(urlData.pathname);
    if (userData !== null && userData.username) {
        username = userData.username;
    }
    return username;
}
exports.extractUsername = extractUsername;
function createChannelUrl(channelId) {
    return `https://www.youtube.com/channel/${channelId}`;
}
exports.createChannelUrl = createChannelUrl;
function createUserUrl(username) {
    return `https://www.youtube.com/user/${username}`;
}
exports.createUserUrl = createUserUrl;
function createPlaylistUrl(playlistId) {
    return `https://www.youtube.com/playlist?list=${playlistId}`;
}
exports.createPlaylistUrl = createPlaylistUrl;
function createVideoUrl(videoId) {
    return `https://www.youtube.com/watch?v=${videoId}`;
}
exports.createVideoUrl = createVideoUrl;
function createThumbnails(videoId) {
    return {
        background: imageUrl(videoId, '0'),
        start: imageUrl(videoId, '1'),
        middle: imageUrl(videoId, '2'),
        end: imageUrl(videoId, '3'),
        high: imageUrl(videoId, 'hqdefault'),
        medium: imageUrl(videoId, 'mqdefault'),
        normal: imageUrl(videoId, 'default'),
        hd: imageUrl(videoId, 'maxresdefault'),
        sd: imageUrl(videoId, 'sddefault')
    };
}
exports.createThumbnails = createThumbnails;
;
function imageUrl(videoId, image) {
    return `https://i1.ytimg.com/vi/${videoId}/${image}.jpg`;
}
function makeAbsolute(path, host = 'https://youtube.com') {
    if (host.endsWith('/'))
        host = host.slice(0, -1);
    return path.startsWith('/') ? (host + path) : path;
}
exports.makeAbsolute = makeAbsolute;
function scrubString(text) {
    return sanitize(text);
}
exports.scrubString = scrubString;
function scrubObject(target) {
    let result = {};
    Object.keys(target).forEach(key => {
        let val = target[key];
        result[key] = typeof val === 'string' ? scrubString(val).trim() : val;
    });
    return result;
}
exports.scrubObject = scrubObject;
function template(format) {
    function customTemplate(data) {
        data = scrubObject(data);
        return customTemplate.lodashTemplate(data);
    }
    customTemplate.format = format;
    customTemplate.lodashTemplate = lodashTemplate(format);
    return customTemplate;
}
exports.template = template;
