// pully/dist/benchmark.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const ytdl = require("ytdl-core");
const ffmpeg = require("fluent-ffmpeg");
const deleteDir = require('rimraf').sync;
const benchmark_1 = require("benchmark");
const _1 = require(".");
const extraSmallVideo = 'https://www.youtube.com/watch?v=oVXg7Grp1W8'; // 22s with music (9MB)
const smallVideo = 'https://www.youtube.com/watch?v=LXb3EKWsInQ'; // Costa Rica 4K, 5:14 (169MB)
const longVideo = 'https://www.youtube.com/watch?v=aqz-KE-bpKQ'; // Big Buck Bunny, 10:34 4K (281 MB)
const videoToDownload = smallVideo;
let downloadSize = 0;
const modeSuite = new benchmark_1.Suite('Download Mode');
modeSuite.on('cycle', function (event) {
    let bm = event.target;
    let opsPerMin = formatOpsSpeed(bm);
    let downloadSpeed = formatDownloadSpeed(bm);
    let size = bm.stats.sample.length;
    console.log(`${bm.name}: [${downloadSpeed}] ${opsPerMin} \xb1${bm.stats.rme.toFixed(2)}% (${size} run${size == 1 ? '' : 's'} sampled)`);
});
modeSuite
    .add(downloadModeBenchmarkFor(_1.DownloadMode.Merge))
    .add(downloadModeBenchmarkFor(_1.DownloadMode.Sequential))
    .add(downloadModeBenchmarkFor(_1.DownloadMode.Parallel))
    .add(ytldCorBenchmark())
    .run({ async: false });
function downloadModeBenchmarkFor(mode) {
    let pully;
    let i = 0;
    let downloadPath;
    deleteDir(path_1.resolve(__dirname, `../temp/${mode}/*`));
    return {
        name: mode,
        defer: true,
        setup: () => {
            pully = new _1.Pully({
                dir: `./temp/${mode}`,
                preset: _1.Presets.HD,
                template: `output_${i++}`,
                mode
            });
        },
        fn: function (p) {
            return __awaiter(this, void 0, void 0, function* () {
                const result = yield pully.download(videoToDownload);
                downloadPath = result.path;
                downloadSize = result.format.downloadSize;
                p.resolve();
            });
        },
        teardown: () => {
            //deleteFile(downloadPath);
        }
    };
}
function ytldCorBenchmark() {
    const audioOutput = path_1.resolve(__dirname, '../temp/ytdl-core/sound.mp4');
    let i = 0;
    deleteDir(path_1.resolve(__dirname, `../temp/ytdl-core/*`));
    return {
        name: 'ytdl-core',
        defer: true,
        setup: () => {
        },
        fn: function (p) {
            return __awaiter(this, void 0, void 0, function* () {
                const mainOutput = path_1.resolve(__dirname, `../temp/ytdl-core/output_${i++}.mp4`);
                ytdl(videoToDownload, { filter: format => format.itag === '140' })
                    // Write audio to file since ffmpeg supports only one input stream.
                    .pipe(fs_1.createWriteStream(audioOutput))
                    .on('finish', () => {
                    ffmpeg()
                        .input(ytdl(videoToDownload, { filter: format => format.itag === '299' }))
                        .videoCodec('copy')
                        .input(audioOutput)
                        .audioCodec('copy')
                        .save(mainOutput)
                        .on('error', console.error)
                        // .on('progress', progress => {
                        //   process.stdout.write(progress.timemark);
                        // })
                        .on('end', () => {
                        fs_1.unlinkSync(audioOutput);
                        p.resolve();
                    });
                });
            });
        },
        teardown: () => {
            //deleteFile(mainOutput);
        }
    };
}
function formatDownloadSpeed(bm) {
    let bytesPerSecond = downloadSize / bm.stats.mean;
    let humanReadableSpeed = humanByteSize(bytesPerSecond);
    return `${humanReadableSpeed}/sec`;
}
function formatOpsSpeed(bm) {
    let speed = bm.hz;
    let unit = 'sec';
    if (speed < 1) {
        speed *= 60;
        unit = 'min';
    }
    if (speed < 1) {
        speed *= 60;
        unit = 'hour';
    }
    return `${benchmark_1.formatNumber((speed).toFixed(2))} ops/${unit}`;
}
function humanByteSize(bytes) {
    const threshold = 1024;
    if (Math.abs(bytes) < threshold) {
        return `${bytes} B`;
    }
    const units = ['KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
    let u = -1;
    do {
        bytes /= threshold;
        ++u;
    } while (Math.abs(bytes) >= threshold && u < units.length - 1);
    return `${bytes.toFixed(1)} ${units[u]}`;
}
