// pully/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const events_1 = require("events");
const debug = require("debug");
const pully_core_1 = require("pully-core");
const models_1 = require("./lib/models");
exports.DownloadMode = models_1.DownloadMode;
const download_1 = require("./lib/download");
const presets_1 = require("./lib/presets");
exports.Presets = presets_1.Presets;
const DEFAULT_TEMPLATE = '${videoTitle}__${channelName}';
const log = debug('pully:index');
class Pully extends events_1.EventEmitter {
    constructor(_config) {
        super();
        this._config = _config;
        this._presets = {};
        this._config = this._config || {};
        this._registerPresets(presets_1.DefaultPresets)._registerPresets(this._config.additionalPresets);
    }
    query(url) {
        return __awaiter(this, void 0, void 0, function* () {
            let results = yield pully_core_1.query(url);
            this.emit('query', results);
            return results;
        });
    }
    download(input, preset) {
        return __awaiter(this, void 0, void 0, function* () {
            const globalStart = Date.now();
            if (typeof input === 'string') {
                input = {
                    url: input,
                    preset
                };
            }
            let options = yield this._formatConfig(input);
            log(`Download started (strategy: ${options.mode})...`);
            this.emit('downloadstarted', options);
            let dlPromise = new download_1.Download(options, this).start();
            dlPromise.then(results => {
                results.duration = Date.now() - globalStart;
                if (results.cancelled) {
                    log(`Download cancelled... %o`, results);
                    this.emit('downloadcancelled', results);
                }
                else {
                    log(`Download complete... %o`, results);
                    this.emit('downloadcomplete', results);
                }
            }, err => this.emit('downloadfailed', { options, err }));
            return dlPromise;
        });
    }
    _formatConfig(input) {
        if (!input) {
            log(`No options detected!`);
            return Promise.reject(new Error(`No options detected!`));
        }
        if (!input.url) {
            log(`URL failed validation. No URL specified.`);
            return Promise.reject(new Error(`"${input.url}" is not a valid URL!`));
        }
        let output = {
            url: input.url,
            dir: null,
            preset: null,
            info: input.info || this._config.info || (() => { }),
            template: this._getTemplate(input.template),
            progress: input.progress,
            mode: input.mode || this._config.mode || models_1.DownloadMode.Parallel
        };
        // Resolve the directory...
        let specifiedDir = input.dir || this._config.dir;
        output.dir = specifiedDir ? path_1.resolve(specifiedDir) : null;
        // Resolve the preset...
        let specifiedPresetName = input.preset || this._config.preset || presets_1.Presets.HD;
        if (typeof specifiedPresetName !== 'string') {
            output.preset = specifiedPresetName;
        }
        else {
            let specifiedPreset = this._presets[specifiedPresetName];
            if (!specifiedPreset) {
                log(`No preset "${specifiedPresetName}"!`);
                let err = new Error(`No preset "${specifiedPresetName}"!`);
                this.emit('downloadfailed', { options: input, err });
                throw err;
            }
            output.preset = specifiedPreset;
        }
        return Promise.resolve(output);
    }
    _registerPresets(presets) {
        if (!presets)
            return this;
        log(`Registering ${presets.length} presets...`);
        presets.forEach(preset => this._presets[preset.name] = presets_1.prepPreset(preset));
        return this;
    }
    _getTemplate(localTemplate) {
        for (let format of [localTemplate, this._config.template, DEFAULT_TEMPLATE]) {
            if (!format)
                continue;
            if (typeof format === 'string') {
                return pully_core_1.template(format);
            }
            else {
                let formatFn = format;
                return (data) => formatFn(pully_core_1.scrubObject(data));
            }
        }
    }
}
exports.Pully = Pully;
