// pully/dist/lib/presets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function fpsFilter(format, options) {
    return !options.maxFps || format.fps <= options.maxFps;
}
function resolutionFilter(format, options) {
    return !options.maxResolution || format.resolution <= options.maxResolution;
}
function audioBitrateFilter(format, options) {
    return !options.maxAudioBitrate || format.audioBitrate <= options.maxAudioBitrate;
}
function fpsSort(a, b) {
    return (b.fps || 0) - (a.fps || 0);
}
function resolutionSort(a, b) {
    return (b.resolution || 0) - (a.resolution || 0);
}
function audioBitrateSort(a, b) {
    return (a.audioBitrate || 0) - (b.audioBitrate || 0);
}
function extendPreset(base, overrides) {
    return Object.assign({}, base, overrides);
}
const baseVideoPreset = {
    name: null,
    maxResolution: Number.MAX_SAFE_INTEGER,
    maxFps: 60,
    videoFilters: [resolutionFilter, fpsFilter],
    videoSort: [resolutionSort, fpsSort],
    audioSort: [audioBitrateSort],
    maxAudioBitrate: 128
};
exports.baseAudioPreset = {
    name: null,
    audioFilters: [audioBitrateFilter],
    audioSort: [audioBitrateSort],
    maxAudioBitrate: 128
};
exports.maxPreset = {
    name: 'max',
    maxFps: Number.MAX_SAFE_INTEGER,
    maxResolution: Number.MAX_SAFE_INTEGER,
    maxAudioBitrate: Number.MAX_SAFE_INTEGER
};
const fourKPreset = {
    name: '4k',
    maxResolution: 2160
};
const twoKPreset = {
    name: '2k',
    maxResolution: 1440
};
const hdPreset = {
    name: 'hd',
    maxResolution: 1080
};
const sdPreset = {
    name: 'sd',
    maxResolution: 720
};
const ldPreset = {
    name: 'ld',
    maxResolution: 480
};
const hfrPreset = {
    name: 'hfr',
    maxFps: Number.MAX_SAFE_INTEGER,
    maxResolution: Number.MAX_SAFE_INTEGER,
    videoSort: [fpsSort, resolutionSort]
};
const mp3Preset = {
    name: 'mp3',
    outputFormat: 'mp3',
    maxAudioBitrate: Number.MAX_SAFE_INTEGER
};
function prepPreset(preset) {
    if (!preset.maxResolution && !preset.maxAudioBitrate) {
        throw new Error(`Custom presets must have at least a maxAudioBitrate or maxResolution`);
    }
    if (!preset.maxResolution) {
        preset = extendPreset(exports.baseAudioPreset, preset);
    }
    else {
        preset = extendPreset(baseVideoPreset, preset);
    }
    return preset;
}
exports.prepPreset = prepPreset;
exports.DefaultPresets = [exports.maxPreset, fourKPreset, twoKPreset, hdPreset, sdPreset, ldPreset, hfrPreset, mp3Preset];
exports.Presets = {
    Max: exports.maxPreset.name,
    FourK: fourKPreset.name,
    TwoK: twoKPreset.name,
    HD: hdPreset.name,
    SD: sdPreset.name,
    LD: ldPreset.name,
    HFR: hfrPreset.name,
    MP3: mp3Preset.name
};
